go/arrow/compute/internal/kernel.go

```diff
@@ -0,0 +1,508 @@
+type MemAlloc int8
+
+const (
+	// For data types that support pre-allocation (fixed-width), the
+	// kernel expects to be provided a pre-allocated buffer to write into.
+	// Non-fixed-width types must always allocate their own buffers.
+	// The allocation is made for the same length as the execution batch,
+	// so vector kernels yielding differently sized outputs should not
+	// use this.
+	//
+	// It is valid for the data to not be preallocated but the validity
+	// bitmap is (or is computed using intersection).
+	//
+	// For variable-size output types like Binary or String, or for nested
+	// types, this option has no effect.
+	MemPrealloc MemAlloc = iota
+	// The kernel is responsible for allocating its own data buffer
+	// for fixed-width output types.
+	MemNoPrealloc
+)
+
+type KernelState any
+
+// KernelInitArgs are the arguments required to initialize an Kernel's
+// state using the input types and any options.
+type KernelInitArgs struct {
+	Kernel Kernel
+	Inputs []arrow.DataType
+	// Options are opaque and specific to the Kernel being initialized,
+	// may be nil if the kernel doesn't require options.
+	Options any
+}
+
+// KernelInitFn is any function that receives a KernelCtx and initialization
+// arguments and returns the initialized state or an error.
+type KernelInitFn = func(*KernelCtx, KernelInitArgs) (KernelState, error)
+
+// KernelSignature holds the input and output types for a kernel.
+//
+// Variable argument functions with a minimum of N arguments should pass
+// up to N input types to be used to validate for invocation. The first
+// N-1 types will be matched against the first N-1 arguments and the last
+// type will be matched against the remaining arguments.
+type KernelSignature struct {
+	InputTypes []InputType
+	OutType    OutputType
+	IsVarArgs  bool
+
+	// store the hashcode after it is computed so we don't
+	// need to recompute it
+	hashCode uint64
+}
+
+func (k KernelSignature) String() string {
+	var b strings.Builder
+	if k.IsVarArgs {
+		b.WriteString("varargs[")
+	} else {
+		b.WriteByte('(')
+	}
+
+	for i, t := range k.InputTypes {
+		if i != 0 {
+			b.WriteString(", ")
+		}
+		b.WriteString(t.String())
+	}
+	if k.IsVarArgs {
+		b.WriteString("*]")
+	} else {
+		b.WriteByte(')')
+	}
+
+	b.WriteString(" -> ")
+	b.WriteString(k.OutType.String())
+	return b.String()
+}
+
+func (k KernelSignature) Equals(other KernelSignature) bool {
+	if k.IsVarArgs != other.IsVarArgs {
+		return false
+	}
+
+	return slices.EqualFunc(k.InputTypes, other.InputTypes, func(e1, e2 InputType) bool {
+		return e1.Equals(&e2)
+	})
+}
+
+func (k *KernelSignature) Hash() uint64 {
+	if k.hashCode != 0 {
+		return k.hashCode
+	}
+
+	var h maphash.Hash
+	h.SetSeed(hashSeed)
+	result := h.Sum64()
+	for _, typ := range k.InputTypes {
+		result = HashCombine(result, typ.Hash())
+	}
+	k.hashCode = result
+	return result
+}
+
+func (k KernelSignature) MatchesInputs(types []arrow.DataType) bool {
+	switch k.IsVarArgs {
+	case true:
```

Comment:
Don't you still want to validate that `types` has at least enough types to match all the non-varargs types?

Reply:
ah, fair. i should do a `len(types) >= (len(k.InputTypes)-1)`